### PR TITLE
Strip auth headers before proxying, add auth headers on upgrade requests, remove auth challenges on responses

### PR DESCRIPTION
**Pull request by liggitt**
We should never leak auth credentials from the request to <openshift>/api through to <proxiedkube>/api

We need to correctly send configured auth when handling upgrade requests ourselves

### CODE REVIEW

**Comment by liggitt** [comment]
@ncdc @derekwaynecarr

File: pkg/util/httpproxy/upgradeawareproxy.go [hunk 1]
```diff
@@ -172,6 +188,7 @@ func (p *UpgradeAwareSingleHostReverseProxy) serveUpgrade(w http.ResponseWriter,
 	done := make(chan struct{}, 2)
 
 	go func() {
+		// TODO: strip any WWW-Authenticate challenge headers if the initial response is a 401
```
Comment (ncdc) [inline pkg/util/httpproxy/upgradeawareproxy.go]: This comment belongs in the 2nd copy goroutine (io.Copy(requestHijackedConn, backendConn))

Comment (ncdc) [inline pkg/util/httpproxy/upgradeawareproxy.go]: The only way I think you could do that is if you call http.ReadResponse(backendConn, req), check for 401, and if it's not 401, write the response to requestHijackedConn and then start the io.Copy

File: pkg/util/httpproxy/upgradeawareproxy.go [hunk 2]
```diff
@@ -183,6 +200,8 @@ func (p *UpgradeAwareSingleHostReverseProxy) serveUpgrade(w http.ResponseWriter,
 	}()
 
 	go func() {
+		// TODO: Intercept a 401 response, log, and transform into a 500 internal server error
```
Comment (ncdc) [inline pkg/util/httpproxy/upgradeawareproxy.go]: Sorry to make you move this again, but it really belongs after the request is written to the backendConn but before the goroutines are started

**Comment by liggitt** [comment]
@derekwaynecarr @ncdc this is working for me

**Comment by smarterclayton** [comment]
Will merge after Derek tries it.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**